src/idaplugin/idaplugin.cpp

```diff
@@ -40,16 +40,23 @@ void killDecompilation()
 
 		if (decompInfo.decompPid)
 		{
-#if defined(OS_WINDOWS)
+			int rc = 0;
+			if (check_process_exit(decompInfo.hDecomp, &rc, 0) != 0)
+				term_process(decompInfo.hDecomp);
+/*#if defined(OS_WINDOWS)
```

Comment:
Well, I guess this is the main issue of this issue :D
I was not aware that the current solution is leaking, but I must admit I wasn't really checking. The problem I tried to solve is to kill not only the spawned process, but the whole process tree. I experimented with IDA SDK's `check_process_exit()` and `term_process()`, but it looks like it terminates only the originally started process - i.e. `retdec-decompiler.py` script. The big problem is, that this script spawns other processes (fortunately only one at the time). So if we kill it from IDA using SDK function, processes like `retdec-bin2llvmir` and `retdec-llvmir2hll` keep running. And they can take a very long time on some inputs. And if user triggers several such long decompilations (and kills them), they stack.

This is caused by RetDec's design (it is not possible to used it like a library at the moment) and a whole IDA plugin is kind of a hack that tries to keep it together.

It would be great, if we could come up with a solution that solves both problems: 1) kills an entire process tree, and 2) does not leak. It looks like you have quite an insight into these things (more than I do), so maybe you will know the answer. But right now, I don't think it would be a step forward to merge this solution solving leaks, but leaving potentially endless processes running on user machines.

Reply:
I did not realize about the process tree - yes now I understand why you have the special OS-specific code in there.

As far as I know, the process tree will be terminated automatically only if a job is created.  CreateJobObject / AssignProcessToJobObject / SetInformationJobObject with JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE / TerminateJobObject.  Other hack like solutions exist like DEBUG_PROCESS and DEBUG_ONLY_THIS_PROCESS WaitForDebugEvent/ContinueDebugEvent taking advantage of a side effect.

A nice solution also is to make the child processes use stdin pipe and let the pipe closing automatically terminate but of course this works only for an IO driven process.